core/server/common/src/main/java/alluxio/master/journal/raft/RaftSnapshotManager.java

```diff
@@ -0,0 +1,310 @@
+  private final Map<InetSocketAddress, RaftJournalServiceClient> mClients;
+
+  private volatile long mLastSnapshotDownloadDurationMs = -1;
+  private volatile long mLastSnapshotDownloadSize = -1;
+  private volatile long mLastSnapshotDownloadDiskSize = -1;
+
+  @Nullable
+  private CompletableFuture<Long> mDownloadFuture = null;
+
+  RaftSnapshotManager(SnapshotDirStateMachineStorage storage) {
+    mStorage = storage;
+
+    InetSocketAddress localAddress = NetworkAddressUtils.getConnectAddress(
+        NetworkAddressUtils.ServiceType.MASTER_RPC, Configuration.global());
+    mClients = ConfigurationUtils.getMasterRpcAddresses(Configuration.global()).stream()
+        .filter(address -> !address.equals(localAddress))
+        .collect(Collectors.toMap(Function.identity(), address -> {
+          MasterSelectionPolicy selection = MasterSelectionPolicy.Factory.specifiedMaster(address);
+          int numTries = 10;
+          int sleep = Math.max(1, mRequestInfoTimeout / numTries);
+          // try to connect to other master once per second for until request info timeout
+          Supplier<RetryPolicy> retry = () -> new ExponentialBackoffRetry(sleep, sleep, numTries);
+          return new RaftJournalServiceClient(selection, retry);
+        }));
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_DOWNLOAD_DURATION_MS.getName(),
+        () -> mLastSnapshotDownloadDurationMs);
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_DOWNLOAD_SIZE.getName(),
+        () -> mLastSnapshotDownloadSize);
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_DOWNLOAD_DISK_SIZE.getName(),
+        () -> mLastSnapshotDownloadDiskSize);
+  }
+
+  /**
+   * Waits synchronously for the download attempt to be complete.
+   * @return the result of the download attempt, or {@link RaftLog#INVALID_LOG_INDEX} if no
+   * attempt is underway
+   */
+  public long waitForAttemptToComplete() {
+    if (mDownloadFuture == null) {
+      return RaftLog.INVALID_LOG_INDEX;
+    }
+    mDownloadFuture.join();
+    // this is to make sure that mDownloadFuture gets reset to null
+    return downloadSnapshotFromOtherMasters();
+  }
+
+  /**
+   * Launches an asynchronous download of the most updated snapshot found on other masters in the
+   * cluster. If the asynchronous download is already in flight, it polls for the results.
+   * @return the log index of the last successful snapshot installation, or -1 if the download is in
+   * flight or has failed.
+   */
+  public long downloadSnapshotFromOtherMasters() {
+    if (mClients.isEmpty()) {
+      SAMPLING_LOG.warn("No followers are present to download a snapshot from");
+      return RaftLog.INVALID_LOG_INDEX;
+    }
+    if (mDownloadFuture == null) {
+      mDownloadFuture = CompletableFuture.supplyAsync(this::core).exceptionally(err -> {
```

Comment:
double check which pool this runs in

Reply:
Changed the pool so it executes in same pool as the `mJournalPool` of `JournalStateMachine`